Mathlib/RingTheory/Smooth/StandardSmooth.lean

```diff
@@ -0,0 +1,229 @@
+/-
+Copyright (c) 2024 Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jung Tao Cheng, Christian Merten, Andrew Yang
+-/
+import Mathlib.Algebra.MvPolynomial.PDeriv
+import Mathlib.LinearAlgebra.Determinant
+import Mathlib.RingTheory.Presentation
+
+/-!
+# Standard smooth algebras
+
+In this file we define standard smooth algebras. For this we introduce
+the notion of a `PreSubmersivePresentation`. This is a presentation `P` that has
+fewer relations than generators. More precisely there exists an injective map from `P.rels`
+to `P.vars`. To such a presentation we may associate a jacobian. `P` is then a submersive
+presentation, if its jacobian is invertible.
+
+Finally, a standard smooth algebra is an algebra that admits a submersive presentation.
+
+## Main definitions
+
+All of these are in the `Algebra` namespace. Let `S` be an `R`-algebra.
+
+- `PreSubmersivePresentation`: A `Presentation` of `S` as `R`-algebra, equipped with an injective
+  map `P.map` from `P.rels` to `P.vars`. This map is used to define the differential of a
+  presubmersive presentation.
+
+For a presubmersive presentation `P` of `S` over `R` we make the following definitions:
+
+- `PreSubmersivePresentation.differential`: A linear endomorphism of `P.rels → P.Ring` sending
+  the `j`-th standard basis vector, corresponding to the `j`-th relation, to the vector
+  of partial derivatives of `P.relation j` with respect to the coordinates `P.map i` for
+  `i : P.rels`.
+- `PreSubmersivePresentation.jacobian`: The determinant of `P.differential`.
+- `PreSubmersivePresentation.jacobiMatrix`: If `P.rels` has a `Fintype` instance, we may form
+  the matrix corresponding to `P.differential`. Its determinant is `P.jacobian`.
+- `SubmersivePresentation`: A submersive presentation is a finite, presubmersive presentation `P`
+  with in `S` invertible jacobian.
+
+Furthermore, for algebras we define:
+
+- `IsStandardSmooth`: `S` is `R`-standard smooth if `S` admits a submersive
+  `R`-presentation.
+- `IsStandardSmooth.relativeDimension`: If `S` is `R`-standard smooth this is the dimension
+  of an arbitrary submersive `R`-presentation of `S`. This is independent of the choice
+  of the presentation (TODO, see below).
+- `IsStandardSmoothOfRelativeDimension n`: `S` is `R`-standard smooth of relative dimension `n`
+  if it admits a submersive `R`-presentation of dimension `n`.
+
+Finally, in the `RingHom` namespace we define
+
+- `IsStandardSmooth`: A ring homomorphism `R →+* S` is standard smooth if `S` is standard smooth
```

Comment:
I am pretty sure that doc-gen can't figure out whether this is the algebra or the ringhom version. So I think it still pays of to be explicit about the namespace when there is ambiguity.